### PR TITLE
Natasha/listening duration

### PR DESCRIPTION
**Pull request by natashaarmbrust**
- Adds listening duration to episode model
- Adds listening duration to feed cells, episode cells, bookmark cells, and episode detail header view
- Sends batch updates to backend for listening history durations 
- Removes all UI code for updating based off a change in the episode `isPlaying` flag. The views all update correctly when switching from another view back to that view.  This should integrate well with Drew's KVO branch (since we will be calling setupWithEpisode on every view that observes the episode isPlaying change)
Note: `EpisodeDurationEndpoint` and `current_progress`/`real_duration_written` fields are not merged into master on backend yet.. I will do thorough testing of this once it is but @kevingreer and I have spoken and our formatting should be consistent. 
- Fixes large image bug
- Fixes temp username bug
- Fixes follower following incorrect updating bug
- Adds some more intuitive text for external profile

### CODE REVIEW

**Comment by natashaarmbrust** [comment]
img_0742
img_0743

File: Podcast/EpisodeUtilityButtonBarView.swift [hunk 1]
```diff
@@ -16,6 +16,7 @@ class EpisodeUtilityButtonBarView: UIView {
     var bookmarkButton: BookmarkButton!
     var moreButton: MoreButton!
     var playButton: PlayButton!
+    var slider: EpisodeDurationSliderView!
     var hasBottomLineseparator: Bool = false
```
Comment (mindylou) [inline Podcast/EpisodeUtilityButtonBarView.swift]: capitalize `Separator` for this var and the one below it

File: Podcast/EpisodeUtilityButtonBarView.swift [hunk 2]
```diff
@@ -94,7 +108,13 @@ class EpisodeUtilityButtonBarView: UIView {
...
 
-    func setPlayButtonToState(isPlaying: Bool) {
-        playButton.isSelected = isPlaying
+    func setupWithEpisode(episode: Episode) {
+        playButton.isSelected = episode.isPlaying
+        slider.setSliderProgress(isPlaying: episode.isPlaying, progress: episode.currentProgress)
+        bookmarkButton.isSelected = episode.isBookmarked
+        recommendedButton.setupWithNumber(isSelected: episode.isRecommended, numberOf: episode.numberOfRecommendations)
+        if !slider.isHidden {
+            topLineseparator.isHidden = true
```
Comment (mindylou) [inline Podcast/EpisodeUtilityButtonBarView.swift]: `topLineSeparator.isHidden = !slider.isHidden`

File: Podcast/Player.swift [hunk 3]
```diff
@@ -209,8 +238,10 @@ class Player: NSObject {
     
     func seekTo(_ position: TimeInterval) {
         if let _ = player.currentItem {
+            currentEpisodePercentageListened += abs(getProgress() - currentTimeAt)
```
Comment (mindylou) [inline Podcast/Player.swift]: this is kinda hard to read/understand from a glance and it seems like it's being used a lot, maybe it could be abstracted to a helper function that updates the episode percentage listened?

File: Podcast/PlayerViewController.swift [hunk 4]
```diff
@@ -288,12 +288,18 @@ class PlayerViewController: TabBarAccessoryViewController, PlayerDelegate, Playe
...
+        case .one:
+            Player.sharedInstance.setSpeed(rate: .one_25)
+        case .one_25:
+            Player.sharedInstance.setSpeed(rate: .one_5)
+        case .one_5:
+            Player.sharedInstance.setSpeed(rate: .one_75)
+        case .one_75:
+            Player.sharedInstance.setSpeed(rate: .two)
+        case .two:
+            Player.sharedInstance.setSpeed(rate: .zero_5)
```
Comment (mindylou) [inline Podcast/PlayerViewController.swift:302]: is it gonna be a bad user experience to toggle through like 5-6 speeds to get desired speed? we should discuss with designers

File: Podcast/SaveListeningDurationEndpointRequest.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    init(listeningDurations: [String: ListeningDuration]) {
+        self.listeningDurations = listeningDurations
+        super.init()
+        path = "/history/listening/"
+        httpMethod = .post
+        bodyParameters = [:]
+        listeningDurations.forEach { (id, listeningDuration) in
+            var episodeJSON: JSON = []
+            if let epsiode = Cache.sharedInstance.get(episode: id), epsiode.isDurationWritten {
```
Comment (mindylou) [inline Podcast/SaveListeningDurationEndpointRequest.swift]: episode** lol

**Review by mindylou — CHANGES_REQUESTED** [review]
looks good so far! a few comments

File: Podcast/Player.swift (hunk 3 above)
Comment (natashaarmbrust) [reply Podcast/Player.swift]: Done

File: Podcast/PlayerViewController.swift (hunk 4 above)
Comment (natashaarmbrust) [reply Podcast/PlayerViewController.swift:302]: Yea I agree we should discuss with them. I will keep it for now but its something we should get design approval / suggestions for.

File: Podcast/SaveListeningDurationEndpointRequest.swift (hunk 5 above)
Comment (natashaarmbrust) [reply Podcast/SaveListeningDurationEndpointRequest.swift]: Done

File: Podcast/EpisodeUtilityButtonBarView.swift (hunk 2 above)
Comment (natashaarmbrust) [reply Podcast/EpisodeUtilityButtonBarView.swift]: Done

File: Podcast/EpisodeUtilityButtonBarView.swift (hunk 1 above)
Comment (natashaarmbrust) [reply Podcast/EpisodeUtilityButtonBarView.swift]: Done

File: Podcast/BookmarkTableViewCell.swift [hunk 6]
```diff
@@ -163,6 +173,7 @@ class BookmarkTableViewCell: UITableViewCell {
     }
     
     func setPlayButtonToState(isPlaying: Bool) {
+        //TODO: change slider value + probably remove with key value observation
```
Comment (doctor3w) [inline Podcast/BookmarkTableViewCell.swift]: No definitely want to leave the state stuff. This is what would be called by the setup method from KVO

File: Podcast/EpisodeTableViewCell.swift [hunk 7]
```diff
@@ -53,10 +53,6 @@ class EpisodeTableViewCell: UITableViewCell, EpisodeSubjectViewDelegate {
         episodeSubjectView.episodeUtilityButtonBarView.setBookmarkButtonToState(isBookmarked: isBookmarked)
     }
     
-    func setPlayButtonToState(isPlaying: Bool) {
```
Comment (doctor3w) [inline Podcast/EpisodeTableViewCell.swift:56]: Why is this removed?

File: Podcast/Player.swift [hunk 8]
```diff
@@ -3,25 +3,40 @@ import UIKit
 import AVFoundation
 import MediaPlayer
 
+class ListeningDuration {
```
Comment (doctor3w) [inline Podcast/Player.swift:6]: Use a struct instead of a class here

**Review by doctor3w — APPROVED** [review]
LGTM

**Review by mindylou — APPROVED** [review]

File: Podcast/Player.swift (hunk 8 above)
Comment (natashaarmbrust) [reply Podcast/Player.swift:6]: Actually I shouldn't do this because I end up updating some of ListeningDurations variables. You can't mutate fields in a stuct without adding extra mutable code that is ugly and unnecessary.

File: Podcast/EpisodeTableViewCell.swift (hunk 7 above)
Comment (natashaarmbrust) [reply Podcast/EpisodeTableViewCell.swift:56]: addressed .. I removed it because we now need to update the slider value as well so we need more than just the isPlaying boolean.
